Fix complaint about nil IP address on restore of cilium_host

christarazi opened the pull request:
- daemon/cmd: Fix complaint about nil IP address on restore of cilium_host
- pkg/datapath/loader: Export SetupBaseDevice
- daemon/cmd: Add new privileged test suite
- pkg/netns: Expand godoc on ReplaceNetNSWithName

Fixes: https://github.com/cilium/cilium/issues/18118

christarazi committed:
daemon/cmd: Add new privileged test suite

This test suite can be used to unit test the daemon code under
daemon/cmd/daemon.go. As a first attempt, sanity check tests for
removing old router (cilium_host) state have been added. This is
intended not to be exhaustive for now.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon/cmd: Fix complaint about nil IP address on restore of cilium_host

Previously, the msg "Failed to remove old router IPs (restored IP:
<nil>) from cilium_host. Manual intervention is required to remove all
other old IPs." would pollute the logs for no good reason. This commit
prevents this log from being printed by avoiding the condition. The
condition is when the cilium_host device has been removed and there are
no IPs detected to restore.

The following commit will add unit tests to verify this behavior.

Fixes: fcd00390c30 ("daemon, node: Remove old, discarded router IPs from `cilium_host`")

Signed-off-by: Chris Tarazi <chris@isovalent.com>

ldelossa commented on daemon/cmd/daemon.go:
```diff
@@ -324,7 +324,9 @@ func (d *Daemon) restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
 func removeOldRouterState(ipv6 bool, restoredIP net.IP) error {
 	l, err := netlink.LinkByName(defaults.HostDevice)
-	if err != nil {
+	if _, linkNotFoundErr := err.(netlink.LinkNotFoundError); linkNotFoundErr && restoredIP == nil {
```
Can `errors.As()` be used here to simplify this?

christarazi replied:
Unfortunately, this is not possible with how netlink [exposes](https://github.com/vishvananda/netlink/blob/c94808a88ba452844600de07ac14ed0ee9badca7/link.go#L1346) `LinkNotFoundError`. I [mentioned](https://github.com/vishvananda/netlink/issues/325#issuecomment-1201609182) this to the maintainer of netlink, and they fairly said to "submit a PR" essentially. :)

ldelossa replied:
cool :)

christarazi replied:
@pippolo84 Tested my assumptions and asserted that `errors.As()` should still work. I was able to confirm it too. Pushed. Thanks both!

ldelossa reviewed (approved):
LGTM, a small note on whether errors.As maybe applicable.

pippolo84 commented on daemon/cmd/daemon_privileged_test.go:
```diff
@@ -0,0 +1,69 @@
+	assert.NotNil(t, netns0)
+	cleanup = func() {
+		cleanup1()
+		netns0.Close()
```
This way, in the cleanup, the deletion of the link device will be completed before the network namespace closing. Is this done on purpose (or is this irrelevant)?

christarazi replied:
Yes I intended for the device to be deleted and then the entire namespace to be deleted. Is that a concern?

pippolo84 replied:
Nope, that's fine! :+1:
I asked just because it caught my eye that this is not the usual LIFO cleanup order that we would get using two separate calls to `defer` or to `t.Cleanup`. Just wanted to be sure I got your intention right.

pippolo84 reviewed (approved):
LGTM, one question left inline.

christarazi committed:
pkg/netns: Expand godoc on ReplaceNetNSWithName

It is useful to have an example usage of how ReplaceNetNSWithName() is
typically used so developers don't have to research how to use network
namespaces in Go again.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
pkg/datapath/loader: Export SetupBaseDevice

This is useful for the upcoming new privileged daemon test suite.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
